patches/0002-Add-crypto-backend-foundation.patch

```diff
@@ -741,8 +877,218 @@ index d7f082c9ee1e04..14a85fbf1bd465 100644
  	return hkdf.Extract(c.hash.New, newSecret, currentSecret)
  }
  
+diff --git a/src/crypto/tls/prf.go b/src/crypto/tls/prf.go
+index 20bac96e86703b..c47391f684dc6f 100644
+--- a/src/crypto/tls/prf.go
++++ b/src/crypto/tls/prf.go
+@@ -7,6 +7,7 @@ package tls
+ import (
+ 	"crypto"
+ 	"crypto/hmac"
++	boring "crypto/internal/backend"
+ 	"crypto/md5"
+ 	"crypto/sha1"
+ 	"crypto/sha256"
+@@ -45,7 +46,13 @@ func pHash(result, secret, seed []byte, hash func() hash.Hash) {
+ }
+ 
+ // prf10 implements the TLS 1.0 pseudo-random function, as defined in RFC 2246, Section 5.
+-func prf10(result, secret, label, seed []byte) {
++func prf10(result, secret, label, seed []byte) error {
++	if boring.Enabled && boring.SupportsTLS1PRF() && boring.SupportsHash(crypto.MD5SHA1) {
++		if err := boring.TLS1PRF(result, secret, label, seed, nil); err != nil {
++			return fmt.Errorf("crypto/tls: prf10: %v", err)
++		}
++		return nil
++	}
+ 	hashSHA1 := sha1.New
+ 	hashMD5 := md5.New
+ 
+@@ -61,16 +68,24 @@ func prf10(result, secret, label, seed []byte) {
+ 	for i, b := range result2 {
+ 		result[i] ^= b
+ 	}
++	return nil
+ }
+ 
+ // prf12 implements the TLS 1.2 pseudo-random function, as defined in RFC 5246, Section 5.
+-func prf12(hashFunc func() hash.Hash) func(result, secret, label, seed []byte) {
+-	return func(result, secret, label, seed []byte) {
++func prf12(h crypto.Hash, hashFunc func() hash.Hash) func(result, secret, label, seed []byte) error {
++	return func(result, secret, label, seed []byte) error {
++		if boring.Enabled && boring.SupportsTLS1PRF() && boring.SupportsHash(h) {
++			if err := boring.TLS1PRF(result, secret, label, seed, hashFunc); err != nil {
++				return fmt.Errorf("crypto/tls: prf12: %v", err)
++			}
++			return nil
++		}
+ 		labelAndSeed := make([]byte, len(label)+len(seed))
+ 		copy(labelAndSeed, label)
+ 		copy(labelAndSeed[len(label):], seed)
+ 
+ 		pHash(result, secret, labelAndSeed, hashFunc)
++		return nil
+ 	}
+ }
+ 
+@@ -85,56 +100,64 @@ var keyExpansionLabel = []byte("key expansion")
+ var clientFinishedLabel = []byte("client finished")
+ var serverFinishedLabel = []byte("server finished")
+ 
+-func prfAndHashForVersion(version uint16, suite *cipherSuite) (func(result, secret, label, seed []byte), crypto.Hash) {
++func prfAndHashForVersion(version uint16, suite *cipherSuite) (func(result, secret, label, seed []byte) error, crypto.Hash) {
+ 	switch version {
+ 	case VersionTLS10, VersionTLS11:
+ 		return prf10, crypto.Hash(0)
+ 	case VersionTLS12:
+ 		if suite.flags&suiteSHA384 != 0 {
+-			return prf12(sha512.New384), crypto.SHA384
++			h := crypto.SHA384
++			return prf12(h, sha512.New384), h
+ 		}
```

Comment:
Why is `h` pulled out rather than passing directly?

Reply:
To avoid having to duplicate the identifier, like in here:

```go
return prf12(crypto.SHA384, sha512.New384), crypto.SHA384
```

Reply:
Ha, I somehow missed that `h` was being used twice. 👍